README
BFD (Brute Force Detection) 1.2 [bfdr-fx.org]
 Copyright (C) 1999-2008, R-fx Networks <proj@r-fx.org>
 Copyright (C) 2008, Ryan MacDonald <ryan@r-fx.org>

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

1) Introduction:
BFD is a modular shell script for parsing application logs and checking for
authentication failures. It does this using a rules system where application
specific options are stored including regular expressions for each unique auth
format. The regular expressions are parsed against logs using the 'sed' tool
(stream editor) which allows for vastly superior performance as opposed to
perl based regular expressions or using other tools such as grep/egrep.

Although some will argue that perl based stream processing is far more flexible
than unix tools, I argue that perl is a bloated framework to depend upon and unix
tools such as sed are universal across *nix. Further, perl is only more flexible
with regular expression based stream 'editing', such as in place editing etc..,
if you are only processing a data stream (as bfd does) then sed wins hands down.

In addition to the benefits of parsing logs in a single stream, BFD also uses a
log tracking system so logs are only parsed from the point which they were last
read. This greatly assists in extending the performance of BFD even further as
we are not constantly reading the same log data. The log tracking system is
compatible with syslog/logrotate style log rotations which allows it to detect
when rotations have happened and grab log tails from both the new log file and
the rotated log file.

You can also leverage BFD to block attackers using any number of tools such as
APF, Shorewall, raw iptables, ip route or execute any custom command. There is 
also a fully customizable e-mail alerting system with an e-mail template that
is well suited for every day use or you can open it up and modify it.

The attacker tracking in BFD is handled using simple flat text files that are
size-controlled to prevent space constraints over time, ideal for disk less
devices. There is also an attack pool where trending data is stored on all hosts
hat have been blocked including which rule the block was triggered by.

In the execution process, there is simply a cron job that executes BFD once
every 3 minutes by default. An embedded lock file system makes sure that no
two instances ever run at the same time, preventing messy and potentially load
heavy results. The cronjob can be run more frequently for those that desire it
and doing so will not cause any performance issues (no less than once a minute).

2) Installation:
The included install.sh script does as it says, it will install bfd to the
'/usr/local/bfd' path and place a 3-minute cronjob in '/etc/cron.d/bfd'. The
setup script will also execute an included 'importconf' script if you have a
previous version of bfd installed, which will import your previous settings.

Although it is supported to customize the default installation layout of BFD,
it is not something I recommend or have done much testing with so do so as
you see fit :-)

3) Configuration:
The configuration file for BFD is located at '/usr/local/bfd/conf.bfd', it is
very straight forward and the comments explain themself as to what each option
is for. The most important option is the TRIG="" value in conf.bfd as this
controls how many failed logins before an address is blocked. You should also
make sure to configure an email address and to turn on email alerting to stay
informed on any bans that are placed.

An ignore file is present at '/usr/loca/bfd/ignore.hosts', this is a line
spaced file to place IP Addr's into that you would like to be ignored for
authentication failures. An internal function will attempt to fetch all local
IP's bound on the installed system and ignore any events appearing to source
from those local addresses (but still send an alert).

The application rule files are located under '/usr/local/bfd/rules/' and
they use a simple check feature to auto-enable them self based on the existence
of specific application files. As an example, if the file '/usr/sbin/exim' is
found to exist then the exim rule will be enabled for processing.

You may wish to set unique trigger values for each application and this can be
done by opening the rule in question (i.e: /usr/local/bfd/rules/exim) and set
the TRIG="" value to whatever amount of failed authentications you would like a
ban to be issued at.

If you would like to have BFD place bans via a facility other than the default
configured APF ban command, you can do so by editing the BAN_COMMAND option in
conf.bfd. As an example you can use a generic system route reject to drop hosts
like so:

BAN_COMMAND="route add -host $ATTACK_HOST reject"

4) License:
BFD is developed and supported on a volunteer basis by Ryan MacDonald
[ryan@r-fx.org]

BFD (Brute Force Detection) is distributed under the GNU General Public
License (GPL) without restrictions on usage or redistribution. The BFD
copyright statement, and GNU GPL, "COPYING.GPL" are included in the top-level
directory of the distribution. Credit must be given for derivative works as
required under GNU GPL.

5) Support:
If you require any assistance with BFD you may refer to the R-fx Networks
community forums located at http://forums.rfxnetworks.com. You may also send
an e-mail to support@r-fx.org.

The offical home page for BFD is located at:
http://www.rfxnetworks.com/bfd.php

All bugs or feature requests should be sent to ryan@r-fx.org and please be sure
to include as much information as possible or conceptual ideas of how you think
a new feature should work.



CHANGELOG
1.2:
[Change] changed interpreter definition from /bin/sh to /bin/bash
[Fix] issue with proftpd regexp corrected
[New] added courier rule for courier imap/pop3
[New] added cpanel rule for cpanel/whm/webmail
[New] added vpopmail rule for qmail pop3
[Fix] importconf was creating a broken symlink to old install path backups
[Change] the alert template has been rewritten for mobile friendly e-mail alerts
[Change] tlog now uses stat -c to grab file lengths instead of ls; both use
         st_size to grab file lengths but stat just does it faster
[Change] rewrite of all existing rules using sed + regexp
[New] all variables in conf.bfd have been renamed for better consistency
[New] all attacking hosts are tracked in bfd/tmp/track.attack and file is kept
      at no more than 2500 lines
[Change] removed the use of grep based pattern.* files for the more efficent
         use of sed based regexp checks
[Change] README file has been updated
[New] many changes and cleanups to the internals of bfd, far more than could
      be listed

0.9:
[Change] created apool_rgen function to update attack pool based on IPs
         currently banned and/or removed
[Fix] importconf template has escape errors on backticks
[Fix] modified sendmail rule; error in colum selection - 13 to 10
[Change] modified importconf to merge log tracking data upon updates
[Change] cleaned some repeated operations in bfd main check() function
[New] added importconf script to merge custom conf vars from previous install
[Change] moved exim rule back to standard path and increased trigger to 50
[Change] modified alert function to better handle and not send alerts for
         duplicate events
[Change] modified alert function to send less but more acurate logs with alerts
[New] added sendmail RCPT REJECT rule
[Change] added a test operation to all rules to determine if service exists
         on local system; if not rule is not processed at all
[Fix] altered the ban command execution style to be more compliant with options
      other than apf such as tcpwrappers
[Change] modified alert.bfd template
[Fix] corrected check() loop error that caused bfd to run untill killed on some
      systems
 
0.8:
[Fix] modified rh_imap/rh_pop3 rule to accomidate for events with no hostname
[Change] modified install.sh to setup 644 instead of 755 perms on cronjob
[Fix] modified tlog to better handle rotation events
[Change] small modifications to rules to more implicitly check for ip's
	 opposed to hostnames

0.7:
[Change] revised sshd rule again, reverted to parsing /var/log/secure, no
	 rhost logs are parsed anymore.
[Fix] used sed to remove traces of ipv6 from sshd logs (:::ffff::)
[Change] modified check() with more granular error checking on IP addresses
[Change] modified alert.bfd, added more conditional output

0.6:
[New] pure-ftpd rule added
[New] exim rule added
[Change] changed default permissions on installation files

0.5:
[Fix] modified sshd rule; handle rhost style logs
[Fix] modified imap rule; handle rhost style logs
[Fix] modified pop3 rule; handle rhost style logs
[Change] added various pattern entries to pattern.auth
[Change] modified alert() function, correct null entry log output
(rev:2)
[Change] added definition of module for triggered events in logs

0.4:
[Fix] used variables not being nulled properly between rule execution
[Fix] ip exclude/ignore routine imporperly excluding certian ip's
[Fix] sshd rule improperly handling 'illegal user' notices
[New] added example to check multiple log files [ensim]; apache rule
[Fix] 'stat' command not compatible with debian, replaced with use of 'ls'
(rev:2)
[Fix] LP variable not redefined during loop for vhost logs; apache rule
[Fix] missing '-d' from if operation; apache rule
(rev:3)
[Fix] minor revision made to sshd rule; was not properly parsing logs

0.3:
[New] rh_pop3 bfd-rule
[New] rh_imap bfd-rule
[New] added tlog script; log length  tracker
[New] added get_state(); - lockfile function
[Change] revised rule file formats
[New] added exclude.files; list of files containing hosts to ignore
[New] added local-ip precheck routine; ensure no local ip is banned
[Change] renamed $HOST var to $ATT_HOST
[Change] revised cronjob from 5 to 8 minute execution cycle
[New] added uninstall.sh / copies to $INSPATH
[Change] revised alert.bfd
[New] added apache rule; validates against HTTP AUTH failures in CLF
      apache error_log's

0.2:
[Change] removed deprecated conf.bfd options
[Change] replaced FWRST/FWRULES with BCMD/FWFILE
[Change] restructured logic of check() function
[Change] error checking routines revised
[New] added contents to README file
[Change] revised all comment headers
[Change] exported alert e-mail to alert.bfd file
[New] added host ignore file; ignore.hosts

0.1:
[New] Inital release


uninstall.sh
#!/bin/bash
# BFD 0.4 [bfd@r-fx.org]
###
# Copyright (C) 1999-2008, R-fx Networks <proj@r-fx.org>
# Copyright (C) 2008, Ryan MacDonald <ryan@r-fx.org>
#
#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 2 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the Free Software
#    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
###
#
INSPATH="/usr/local/bfd"
BINPATH="/usr/local/sbin/bfd"
APPN="BFD"

uninstall(){
echo "Remove $APPN from this system; are you sure ?"
echo "Press any key to continue or ^C to abort."
read val

if [ -d "$INSPATH" ]; then
	rm -rf $INSPATH $BINPATH /etc/cron.d/bfd /etc/logrotate.d/bfd /var/log/bfd_log
	echo "$APPN has been uninstalled."
else
	echo "$APPN does not appear to be installed."
fi
}

uninstall


install.sh
#!/bin/bash
# BFD 0.9 [bfd@r-fx.org]
###
# Copyright (C) 1999-2008 R-fx Networks <proj@r-fx.org>
# Copyright (C) 2008, Ryan MacDonald <ryan@r-fx.org>
#
#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 2 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the Free Software
#    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
###
#
INSPATH="/usr/local/bfd"
BINPATH="/usr/local/sbin/bfd"

backup(){
if [ -d "/usr/local/bfd" ]; then
        DVAL=`date +"%d%m%Y-%s"`
	mv /usr/local/bfd/ /usr/local/bfd.bk.$DVAL
	rm -f /usr/local/bfd.bk.last
	ln -s /usr/local/bfd.bk.$DVAL /usr/local/bfd.bk.last
	OBK=1
fi
}

install(){
        rm -rf $INSPATH
        mkdir $INSPATH
	cp logrotate.d.bfd /etc/logrotate.d/bfd
        cp -R files/* $INSPATH
	cp README CHANGELOG COPYING.GPL $INSPATH
	rm -f /etc/cron.daily/bfd
	cp cron.daily /etc/cron.daily/bfd
	chmod 755 /etc/cron.daily/bfd
        chmod 640 $INSPATH/*
        chmod 750 $INSPATH/tlog
        chmod 750 $INSPATH/bfd
	chmod 750 $INSPATH/rules
	chmod 750 $INSPATH/tmp
	chmod 750 $INSPATH/alert.bfd
        ln -fs $INSPATH/bfd $BINPATH
	if [ -f "uninstall.sh" ]; then
		cp uninstall.sh $INSPATH/
		chmod 750 $INSPATH/uninstall.sh
	fi
	if [ -f "cron" ]; then
		cp cron /etc/cron.d/bfd
		chmod 644 /etc/cron.d/bfd
	fi
}

postinfo(){
	echo ".: BFD installed"
	echo "Install path:    $INSPATH"
	echo "Config path:     $INSPATH/conf.bfd"
	echo "Executable path: $BINPATH"
}

if [ -d "$INSPATH" ]; then
	backup	
	install
	postinfo
	./importconf
else
	install
	postinfo
fi
